### PR TITLE
EnC rude edit for local function attributes

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/48636

### CODE REVIEW

**Comment by davidwengier** [comment]
FYI @tmat if you get a chance to take a look before you go, that would be great.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 1]
```diff
@@ -3742,6 +3742,10 @@ private static bool TryMapParameter((SyntaxNode? Node, int Ordinal) parameterKey
             {
                 rudeEdit = RudeEditKind.ChangingLambdaReturnType;
             }
+            else if (!oldLambdaSymbol.GetAttributes().SequenceEqual(newLambdaSymbol.GetAttributes()))
+            {
+                rudeEdit = RudeEditKind.ChangingLambdaAttributes;
+            }
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3748]: These lines of code brought to you by a days worth of "Step Into" 😁

Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3748]: I think this is roughly the right place to report these. However, a few points to consider:
1) we should report the rude edit in addition to the other signature rude edits reported here since they are unrelated changes.
2) I'd move the reporting to a separate method from ReportLambdaSignatureRudeEdits - attributes are not part of lambda signature and the caller makes subsequent decisions based on whether there is a rude edit reported for the actual signature. These decisions shouldn't be affected by attribute changes.
3) We currently report attribute edits based on syntax comparison in [http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/EditAndContinue/CSharpEditAndContinueAnalyzer.cs,287](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/EditAndContinue/CSharpEditAndContinueAnalyzer.cs,287) and using a general "update" rude edit (no need to add a new RudeEditKind. I think we should do the same for local function. Or even better have a method that both places call to report any edits to attributes.
4) We also need to check attributes applied to parameters of a local function.

---
In reply to: [513984575](https://github.com/dotnet/roslyn/pull/49027#discussion_r513984575) [](ancestors = 513984575)

**Review by tmat — CHANGES_REQUESTED** [review]
:clock1:

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 1 above)
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3748]: 1. Makes sense
2. Sure
3. Wrong link? Method attribute rude edits are reported in http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/EditAndContinue/CSharpEditAndContinueAnalyzer.cs,1889 but that doesn't ever get called for method contents, so can't see local function attributes. I can certainly not add a new RudeEditKind of course. I might be missing something here though.
4. Yep, thats why this is a draft, but want to make sure i'm on the right track first.

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs [hunk 2]
```diff
@@ -3849,6 +3849,141 @@ public void MethodUpdate_LambdaParameterRefnessInBody()
                 "Update [public void M(int a) { f((ref int b) => b = 1); }]@10 -> [public void M(int a) { f((out int b) => b = 1); }]@10");
         }
 
+        [Fact]
+        public void MethodUpdate_LocalFunctionAddAttribute()
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:3853]: >MethodUpdate_LocalFunctionAddAttribute [](start = 20, length = 38)

These tests should rather be in StatementEditingTests (near LocalFunctions_InExpressionStatement). After all, local functions are not "top-level". I think the above 2 tests are validating that update to a local function / lambda signature is not treated as top-level update.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 3]
```diff
@@ -3708,6 +3710,58 @@ private static bool TryMapParameter((SyntaxNode? Node, int Ordinal) parameterKey
...
+            var newLambdaSymbol = GetLambdaExpressionSymbol(newModel, newLambda, cancellationToken);
+
+            var reportDiagnostic = false;
+            if (!oldLambdaSymbol.GetAttributes().SequenceEqual(newLambdaSymbol.GetAttributes()) ||
+                !oldLambdaSymbol.GetReturnTypeAttributes().SequenceEqual(newLambdaSymbol.GetReturnTypeAttributes()))
+            {
+                reportDiagnostic = true;
+            }
+
+            // If the old and new signatures are the same then we are safe to compare parameter attributes more easily
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3742]: >are safe to compare parameter attributes more easily [](start = 66, length = 52)

I'd say that we don't need to report rude edits for parameter attributes if the signatures don't match since we already reported rude edit for the signature mismatch

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -3708,6 +3710,58 @@ private static bool TryMapParameter((SyntaxNode? Node, int Ordinal) parameterKey
...
+                    {
+                        reportDiagnostic = true;
+                        break;
+                    }
+                }
+            }
+
+            if (reportDiagnostic)
+            {
+                diagnostics.Add(new RudeEditDiagnostic(
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3757]: This will report "Updating local function will prevent the debugging session from continuing" and squiggle the local function. That's overly generic/imprecise, since some updates to local function are allowed.
It'd be better to squiggle the specific attribute that doesn't match.

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs [hunk 5]
```diff
@@ -3849,6 +3849,141 @@ public void MethodUpdate_LambdaParameterRefnessInBody()
...
+                "Update [void M() { void L([A]int i) { } }]@10 -> [void M() { void L(int i) { } }]@10");
+
+            edits.VerifySemantics(
+                expectedDiagnostics: new[] { Diagnostic(RudeEditKind.Update, "L", CSharpFeaturesResources.local_function) });
+        }
+
+        [Fact]
+        public void MethodUpdate_LocalFunctionParameter_ReorderAttribute()
+        {
+            var src1 = "class C { void M() { void L([A, B]int i) { } } }";
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:3975]: >A, B] [](start = 53, length = 5)

Let's also add a test that changing `[A][B]` to `[A,B]` and vice versa is allowed.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 6]
```diff
@@ -3708,6 +3710,58 @@ private static bool TryMapParameter((SyntaxNode? Node, int Ordinal) parameterKey
             }
         }
 
+        protected virtual void ReportLambdaAttributeRudeEdits(SemanticModel oldModel,
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3713]: >virtual [](start = 18, length = 7)

Can be `private` since it's not overridden by language specific analyzer.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 7]
```diff
@@ -3708,6 +3710,58 @@ private static bool TryMapParameter((SyntaxNode? Node, int Ordinal) parameterKey
...
+            if (!IsNestedFunction(newLambda))
+            {
+                return;
+            }
+
+            var oldLambdaSymbol = GetLambdaExpressionSymbol(oldModel, oldLambda, cancellationToken);
+            var newLambdaSymbol = GetLambdaExpressionSymbol(newModel, newLambda, cancellationToken);
+
+            var reportDiagnostic = false;
+            if (!oldLambdaSymbol.GetAttributes().SequenceEqual(newLambdaSymbol.GetAttributes()) ||
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3736]: >SequenceEqual [](start = 49, length = 13)

`AttributeData` does not implement equality, so I think this compares reference identity which is probably not what we want.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 3 above)
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3742]: Yes, that is what this is doing.. I'll reword the comment.

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs (hunk 5 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:3975]: Changing to syntax means this _won't_ be allowed, but I'll add the test anyway so it won't be forgotten in future if attribute edits are allowed in some fashion.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 4 above)
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3757]: Before I go and write one, is there a helper also that compares sequences and also returns the item that differs?

Or perhaps just squiggling the local function name is sufficient?

Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: I don't remember if we have such a helper. Look around a bit and if you can't find it feel free to add one.
In general we try to be as specific as we can (within reason). Makes the experience much better.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 8]
```diff
@@ -1343,14 +1454,7 @@ internal static new TextSpan GetDiagnosticSpan(SyntaxNode node, EditKind editKin
...
                     var attributeList = (AttributeListSyntax)node;
-                    if (editKind == EditKind.Update)
-                    {
-                        return (attributeList.Target != null) ? attributeList.Target.Span : attributeList.Span;
-                    }
-                    else
-                    {
-                        return attributeList.Span;
-                    }
+                    return attributeList.Span;
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1457]: This change is because otherwise adding `[return: A]` to a local function would just squiggle `return:` which didn't seem nice, so now this method makes no assumptions, and the caller can dictate what gets squiggled.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 9]
```diff
@@ -2869,7 +2973,7 @@ private void ClassifyUpdate(AttributeListSyntax oldNode, AttributeListSyntax new
             {
                 if (!SyntaxFactory.AreEquivalent(oldNode.Target, newNode.Target))
                 {
-                    ReportError(RudeEditKind.Update);
+                    ReportError(RudeEditKind.Update, spanNode: ((SyntaxNode)newNode.Target) ?? newNode);
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:2976]: This is the maintain existing functionality with the change in `GetDiagnosticSpan`

**Comment by davidwengier** [comment]
@tmat I believe all of your feedback has been address, PTAL.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 10]
```diff
@@ -1162,6 +1162,116 @@ internal override bool QueryClauseLambdasTypeEquivalent(SemanticModel oldModel,
...
+            if (!IsLocalFunctionBody(oldLambdaBody) || !IsLocalFunctionBody(newLambdaBody))
+            {
+                return;
+            }
+
+            var newLocalFunction = (LocalFunctionStatementSyntax)GetLambda(newLambdaBody);
+            var oldLocalFunction = (LocalFunctionStatementSyntax)GetLambda(oldLambdaBody);
+
+            var reportDiagnostic = false;
+            if (!AreEquivalentAttributeLists(oldLocalFunction.AttributeLists, newLocalFunction.AttributeLists, out var diagnosticNode))
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1178]: >if [](start = 12, length = 2)

Nit: `var reportDiagnostic = !AreEquivalentAttributeLists(...);`

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 11]
```diff
@@ -1162,6 +1162,116 @@ internal override bool QueryClauseLambdasTypeEquivalent(SemanticModel oldModel,
...
+                }
+            }
+
+            if (reportDiagnostic)
+            {
+                diagnostics.Add(new RudeEditDiagnostic(
+                    RudeEditKind.Update,
+                    GetDiagnosticSpan(diagnosticNode ?? newLocalFunction, EditKind.Update),
+                    newLocalFunction,
+                    new[] { GetDisplayName(newLocalFunction) }));
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1205]: >newLocalFunction [](start = 43, length = 16)

This could also take `diagnosticNode ?? newLocalFunction`, so that the error message reads "Updating 'attribute' ..." rather than "Updating 'local function' ...".

**Comment by davidwengier** [comment]
Have pushed updates for all PR feedback, and for handling type parameter attribute changes.

Working on changing the StatementSyntaxComparer is going to take a bit longer, as there were a bunch of things changed with existing tests, due to edits being different, and it seems there isn't any current test infrastructure for reporting rude diagnostics from the StatementSyntaxComparer, so my trust in my own changes is pretty low :)

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 12]
```diff
@@ -1299,12 +1299,12 @@ public LambdaInfo WithMatch(Match<SyntaxNode> match, SyntaxNode newLambdaBody)
                 {
                     // Body match of a lambda whose body is an expression has the lambda as a root.
                     // The lambda has already been included when enumerating parent body matches.
-                    Debug.Assert(
-                        !map.ContainsKey(pair.Key) ||
-                        pair.Key == lambdaBodyMatch.OldRoot && pair.Value == lambdaBodyMatch.NewRoot && IsLambda(pair.Key) && IsLambda(pair.Value));
-
-                    map[pair.Key] = pair.Value;
-                    reverseMap[pair.Value] = pair.Key;
+                    if (!map.ContainsKey(pair.Key) ||
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1302]: >if [](start = 20, length = 2)

Why is the assertion not true?

File: src/EditorFeatures/CSharpTest/EditAndContinue/CSharpEditAndContinueAnalyzerTests.cs [hunk 13]
```diff
@@ -216,7 +216,6 @@ void M()
         /*<span>*/expr;/*</span>*/
         /*<span>*/int a;/*</span>*/
         F(/*<span>*/(x)/*</span>*/ => x);
-        F(/*<span>*/x/*</span>*/ => x);
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/CSharpEditAndContinueAnalyzerTests.cs:219]: >F(/*<span>*/x/*</span>*/ => x); [](start = 8, length = 31)

Why has this changed?

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 12 above)
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1302]: We're now matching more things about lambda.. So lambdaBodyMatches now has matches for the lambdas parameter list for example.

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 14]
```diff
@@ -6933,6 +6943,182 @@ public void LocalFunction_RemoveAsync()
                 Diagnostic(RudeEditKind.ChangingFromAsynchronousToSynchronous, "local", FeaturesResources.local_function));
         }
 
+        [Fact]
+        public void LocalFunction_AddAttribute()
+        {
+            var src1 = "void L() { }";
+            var src2 = "[A]void L() { }";
+
+            var edits = GetMethodEdits(src1, src2);
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs:6952]: >GetMethodEdits [](start = 24, length = 14)

We should also check rude edits in these tests. That means we need to use GetTopEdits helper since the analyzer only operates on entire tree, not just method body. 
But should be easy enough to wrap each of the source snippets in `class C { void M() { ... } }`.

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 15]
```diff
@@ -1894,7 +1897,7 @@ public void Lambdas_InLambda_ChangeInLambdaSignature()
 
             // changes were made to the outer lambda signature:
             edits.VerifyEdits(
-                "Update [() => { G(x => y); }]@4 -> [q => { G(() => y); }]@4");
+                "Update [() => { G(x => y); }]@4 -> [q => { G(() => y); }]@4", "Insert [q]@4", "Delete [()]@4");
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs:1900]: >", [](start = 76, length = 2)

nit: let's move the Insert on a separate line

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 16]
```diff
@@ -6933,6 +6943,182 @@ public void LocalFunction_RemoveAsync()
...
+
+        [Fact]
+        public void LocalFunction_SplitAttributeLists()
+        {
+            var src1 = "[A, B]void L() { }";
+            var src2 = "[A][B]void L() { }";
+
+            var edits = GetMethodEdits(src1, src2);
+
+            edits.VerifyEdits("Update [[A, B]]@2 -> [[A]]@2", "Insert [[B]]@5", "Insert [B]@6", "Delete [B]@6");
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs:6998]: >" [](start = 62, length = 1)

nit: I think it'd be more readable if each edit was on a separate line

File: src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs [hunk 17]
```diff
@@ -221,6 +221,15 @@ internal enum Label
             GroupClauseLambda,                // tied to parent
             QueryContinuation,                // tied to parent
 
+            TypeParameterList,                // tied to parent
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs:224]: >TypeParameterList [](start = 12, length = 17)

perhaps add a comment above this line that the following are child nodes of local function statement and lambdas

File: src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs [hunk 18]
```diff
@@ -503,6 +520,38 @@ internal static Label Classify(SyntaxKind kind, SyntaxNode? node, out bool isLea
                 case SyntaxKind.AwaitExpression:
                     return Label.AwaitExpression;
 
+                case SyntaxKind.TypeParameterList:
+                    isLeaf = false;
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs:524]: >isLeaf = false; [](start = 20, length = 15)

nit: `isLeaf` is false by default, no need to assign it unless we want to set it to true

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 19]
```diff
@@ -1139,6 +1139,36 @@ internal override bool QueryClauseLambdasTypeEquivalent(SemanticModel oldModel,
...
 
+        protected override void ReportMethodBodySyntaxRudeEditsForLambda(SyntaxNode oldLambda, SyntaxNode newLambda, Match<SyntaxNode> bodyMatch, List<RudeEditDiagnostic> diagnostics)
+        {
+            // We're only interested in local functions here
+            if (!IsLocalFunction(oldLambda))
+            {
+                return;
+            }
+
+            var bodyEditsForLambda = bodyMatch.GetTreeEdits();
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1150]: I don't like this. Ideally this would go in the caller, but there is an issue with this where some of our test cases fail when trying to get the tree edits, so I'm guessing there is a bug in the statement syntax comparer, but I haven't dug into it. The `if (!IsLocalFunction..)` is preventing that test failure for now.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 20]
```diff
@@ -1139,6 +1139,36 @@ internal override bool QueryClauseLambdasTypeEquivalent(SemanticModel oldModel,
...
+            var editMap = BuildEditMap(bodyEditsForLambda);
+            foreach (var edit in bodyEditsForLambda.Edits)
+            {
+                // We are processing edits on the method body that contains the lambda as
+                // things like local function attributes are actually edits in the containing method, not the actual
+                // lambda body.
+                // We only want to consider the edits that are related to the lambda (ie, descendants).
+                if ((edit.OldNode != null && oldLambda.Contains(edit.OldNode)) ||
+                    (edit.NewNode != null && newLambda.Contains(edit.NewNode)))
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1159]: Ideally this can be removed too. I think it would make sense if this method just did syntactic analysis for every method body edit, via ClassifyEdit(), but the bug mentioned above precludes this.

**Comment by davidwengier** [comment]
I've updated this to now use ClassifyEdit to report the rude edits. I like the approach in general, because it will be easy to add type parameter rude edits, and modifier rude edits etc. but there is an issue that I need to look into (and hence have to talk to you about first 😛)

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 21]
```diff
@@ -1259,6 +1261,8 @@ public LambdaInfo WithMatch(Match<SyntaxNode> match, SyntaxNode newLambdaBody)
                         if (newLambdaBody1 != null)
                         {
                             lambdaBodyMatches.Add(ComputeLambdaBodyMatch(oldLambdaBody1, newLambdaBody1, activeNodes, lazyActiveOrMatchedLambdas, diagnostics));
+
+                            ReportMethodBodySyntaxRudeEditsForLambda(GetLambda(oldLambdaBody1), GetLambda(newLambdaBody1), bodyMatch, diagnostics);
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1265]: >GetLambda(oldLambdaBody1) [](start = 69, length = 25)

`oldNode` and `newNode` are the nodes that represent the lambda (see the comment on `TryGetLambdaBodies`: "Returns all lambda bodies of a node representing a lambda").

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 19 above)
Comment (tmat) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1150]: What tests are failing?

---
In reply to: [557054404](https://github.com/dotnet/roslyn/pull/49027#discussion_r557054404) [](ancestors = 557054404)

Comment (davidwengier) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1150]: There are 5 that fail without the if statement, from asserts in StatementSyntaxComparer, for example http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/EditAndContinue/StatementSyntaxComparer.cs,545 throws when it seems a EqualsValueSyntax node. The comment above that assert explains it though - it seems the statement syntax comparer is not necessarily written to be called arbitrarily.

Comment (tmat) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1150]: I'll take a look.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 22]
```diff
@@ -1919,13 +1934,23 @@ public void ClassifyEdit()
...
                 // as long as it represents semantically the same namespace as the one of the original type declaration.
                 ReportError(RudeEditKind.Move);
             }
 
             private void ClassifyReorder(SyntaxNode newNode)
             {
+                if (!_isTopLevelEdit)
+                {
+                    return;
+                }
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1952]: Why do we not report reordering of type parameters, e.g. for local functions?

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 23]
```diff
@@ -1919,13 +1934,23 @@ public void ClassifyEdit()
 
             private void ClassifyMove()
             {
+                if (!_isTopLevelEdit)
+                {
+                    return;
+                }
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1940]: Why is this needed?

File: src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs [hunk 24]
```diff
@@ -89,24 +90,59 @@ private IEnumerable<SyntaxNode> EnumerateRootChildren(SyntaxNode root)
...
+                        {
+                            yield return descendant;
+                        }
+                    }
+                }
             }
-            else
+        }
+
+        private IEnumerable<SyntaxNode> GetRootChildren(SyntaxNode node)
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/StatementSyntaxComparer.cs:113]: >GetRootChildren [](start = 40, length = 15)

Looks good. I'd just rename this method to `EnumerateChildren` - GetRootChildren suggests that node is a root. But it is not expected to be a root.
Also, can we now merge ENumerateNonRootChildren and EnumerateRootChildren? Seems like we have one extra layer there.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 25]
```diff
@@ -1989,113 +2014,125 @@ private void ClassifyReorder(SyntaxNode newNode)
 
             private void ClassifyInsert(SyntaxNode node)
             {
+                // Check edits that are rude as top level or statement edits
                 switch (node.Kind())
                 {
-                    case SyntaxKind.GlobalStatement:
-                        // TODO:
+                    // These are rude when added to methods, or local functions
+                    case SyntaxKind.TypeParameter:
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:2021]: >case SyntaxKind.TypeParameter: [](start = 20, length = 30)

Why is SyntaxKind.Parameter not here?

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 26]
```diff
@@ -1860,6 +1874,7 @@ protected override string GetSuspensionPointDisplayName(SyntaxNode node, EditKin
                 _kind = kind;
                 _span = span;
                 _match = match;
+                _isTopLevelEdit = isTopLevelEdit;
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1877]: >_isTopLevelEdit [](start = 16, length = 15)

I'm not seeing why is this flag necessary. Wouldn't the syntax kinds that are not found in local functions be just ignored?

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 27]
```diff
@@ -6933,6 +6943,288 @@ public void LocalFunction_RemoveAsync()
...
+            // Get top edits so we can validate rude edits
+            edits = GetTopEdits(WrapMethodBodyWithClass(src1), WrapMethodBodyWithClass(src2));
+
+            edits.VerifyRudeDiagnostics(Diagnostic(RudeEditKind.Delete, "L", FeaturesResources.attribute));
+        }
+
+        [Fact]
+        public void LocalFunction_ReturnType_ReorderAttribute()
+        {
+            var src1 = "[return: A, B]int L() { return 1; }";
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs:7111]: >A, B [](start = 33, length = 4)

Changes in attributes should be rude.

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 28]
```diff
@@ -6933,6 +6943,288 @@ public void LocalFunction_RemoveAsync()
...
+            var src2 = "[B, A]void L() { }";
+
+            var edits = GetMethodEdits(src1, src2);
+
+            edits.VerifyEdits("Reorder [B]@6 -> @3");
+
+            // Get top edits so we can validate rude edits
+            edits = GetTopEdits(WrapMethodBodyWithClass(src1), WrapMethodBodyWithClass(src2));
+
+            edits.VerifyRudeDiagnostics();
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs:6995]: >VerifyRudeDiagnostics [](start = 18, length = 21)

I think this needs to be a rude edit.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 26 above)
Comment (davidwengier) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1877]: Yeah, I think this can be removed now, just haven't gone back to check yet. This is from previous attempts where the method body itself was classified, and I needed a way to ignore more node types.

Comment (davidwengier) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1877]: So this is still an issue because the EditClassifier is now seeing edits for nodes that StatementSyntaxComparer cares about, but TopSyntaxComparer doesn't, and it was written for the latter I guess.

1. Change `_isTopLevelEdit` to `_throwOnUnexpectedNode` and have it just control the `throw ExceptionUtilities.UnexpectedValue` calls
2. Check `HasLabel` before throwing unexpected value
3. Add all of the missing syntax kinds to the EditClassifier and just return without doing anything
4. Leave as is

Thoughts? Or have I missed something about when to class the EditClassifier?

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 25 above)
Comment (davidwengier) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:2021]: Existing tests allow inserting parameters to local functions: https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs#L6737-L6763

File: src/EditorFeatures/CSharpTest/EditAndContinue/CSharpEditAndContinueAnalyzerTests.cs (hunk 13 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharpTest/EditAndContinue/CSharpEditAndContinueAnalyzerTests.cs:220]: Fixed

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 26 above)
Comment (tmat) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1877]: I'd check `HasLabel` _before_ calling the classifier for a specific edit.
